File: templates/problems/06_linked_list/03_linked_list_cycle/typescript/linked_list_cycle.test.ts
import { describe, it, expect } from "vitest";
import { buildCyclicList } from "./helpers";
import { hasCycle } from "./linked_list_cycle";

describe("hasCycle", () => {
    it("nil list", () => {
        expect(hasCycle(buildCyclicList([], -1))).toBe(false);
    });

    it("single no cycle", () => {
        expect(hasCycle(buildCyclicList([1], -1))).toBe(false);
    });

    it("single self cycle", () => {
        expect(hasCycle(buildCyclicList([1], 0))).toBe(true);
    });

    it("cycle at head", () => {
        expect(hasCycle(buildCyclicList([3, 2, 0, -4], 0))).toBe(true);
    });

    it("cycle at middle", () => {
        expect(hasCycle(buildCyclicList([3, 2, 0, -4], 1))).toBe(true);
    });

    it("no cycle", () => {
        expect(hasCycle(buildCyclicList([1, 2, 3, 4, 5], -1))).toBe(false);
    });

    it("two nodes cycle", () => {
        expect(hasCycle(buildCyclicList([1, 2], 0))).toBe(true);
    });
});

File: templates/problems/06_linked_list/03_linked_list_cycle/go/linked_list_cycle_test.go
package linked_list_cycle

import "testing"

func TestHasCycle(t *testing.T) {
	tests := []struct {
		name     string
		vals     []int
		cyclePos int // index where tail connects; -1 means no cycle
		want     bool
	}{
		{"nil list", nil, -1, false},
		{"single no cycle", []int{1}, -1, false},
		{"single self cycle", []int{1}, 0, true},
		{"cycle at head", []int{3, 2, 0, -4}, 0, true},
		{"cycle at middle", []int{3, 2, 0, -4}, 1, true},
		{"no cycle", []int{1, 2, 3, 4, 5}, -1, false},
		{"two nodes cycle", []int{1, 2}, 0, true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			head := buildList(tt.vals)
			if tt.cyclePos >= 0 && head != nil {
				// find tail
				tail := head
				for tail.Next != nil {
					tail = tail.Next
				}
				// find cycle target
				target := head
				for i := 0; i < tt.cyclePos; i++ {
					target = target.Next
				}
				tail.Next = target
			}
			got := hasCycle(head)
			if got != tt.want {
				t.Errorf("hasCycle() = %v, want %v", got, tt.want)
			}
		})
	}
}
